src/TropicalGeometry/semiring.jl

```diff
@@ -0,0 +1,504 @@
+
+################################################################################
+#
+#  Types
+#
+################################################################################
+
+# We use T to record whether we are in the min/max case
+# T is either typeof(min) or typeof(max)
+mutable struct TropicalSemiring{T} <: Field
+end
+
+# We use the flag isinf to denote +/- infinity
+# todo: should this be called TropicalNumber instead?
+#   I have no preference, except that it should be consistent with the other libraries,
+#   e.g. what are elements of p-adic number rings called?
+mutable struct TropicalSemiringElem{T} <: FieldElem
+  parent::TropicalSemiring{T}
+  isinf::Bool
+  data::fmpq
+
+  function TropicalSemiringElem(R::TropicalSemiring{T}, isinf::Bool) where {T}
+    @assert isinf
+    z = new{T}()
+    z.isinf = true
+    z.parent = R
+    return z
+  end
+
+  function TropicalSemiringElem(R::TropicalSemiring{T}, x::RingElem) where {T}
+    return new{T}(R, false, x)
+  end
+end
+
+# Type gymnastics
+
+Oscar.elem_type(::Type{TropicalSemiring{T}}) where {T} = TropicalSemiringElem{T}
+
+Oscar.parent_type(::Type{TropicalSemiringElem{T}}) where {T} = TropicalSemiring{T}
+
+################################################################################
+#
+#  Constructors for the tropical semiring
+#
+################################################################################
+
+# Invoke via tropical_semiring()
+tropical_semiring() = TropicalSemiring{typeof(min)}()
+
+# Invoke via tropical_semiring(max)
+tropical_semiring(::typeof(max)) = TropicalSemiring{typeof(max)}()
+
+# Invoke via tropical_semiring(min)
+tropical_semiring(::typeof(min)) = TropicalSemiring{typeof(min)}()
+
+################################################################################
+#
+#  Constructors for elements
+#
+################################################################################
+
+function (R::TropicalSemiring)(u::TropicalSemiringElem)
+  @assert parent(u) === R
+  return u
+end
+
+function (R::TropicalSemiring)(u::RingElem)
+  v = QQ(u)
+  @assert parent(v) === QQ
+  return TropicalSemiringElem(R, v)
+end
+
+function (R::TropicalSemiring)(u::Union{Integer, Rational})
+  return TropicalSemiringElem(R, QQ(u))
+end
+
+inf(R::TropicalSemiring) = TropicalSemiringElem(R, true)
+
+Oscar.zero(T::TropicalSemiring) = inf(T)
+
+Oscar.one(R::TropicalSemiring) = R(zero(QQ))
+
+Oscar.zero(x::TropicalSemiringElem) = zero(parent(x))
+
+(R::TropicalSemiring)() = zero(R)
+
+################################################################################
+#
+#  Basic access
+#
+################################################################################
+
+# The underlying rational number. This is undefined for inf.
+data(x::TropicalSemiringElem) = x.data
+
+# Test if something is inf.
+isinf(x::TropicalSemiringElem) = x.isinf
+
+Oscar.parent(x::TropicalSemiringElem) = x.parent
```

Comment:
If `TropicalSemiring{T}` is a singleton, then `TropicalSemiring{T}()` always returns the same (identical) object and is also efficient.
```suggestion
Oscar.parent(x::TropicalSemiringElem{T}) where T = TropicalSemiring{T}()
```